### ISSUE TITLE
C# Powershell.SDK - Process created by CreateOutOfProcessRunspace never terminates

### ISSUE THREAD

**Issue by SirDester** [PowerShell/PowerShell#23991]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I'm using the latest (7.4.3) version of Powershell SDK.

From c# code I need to open a Powershell Runspace for a 5.1 Environment (because I need to load Sharepoint PsSnapin and execute the commands, and this module is not supported by the Powershell 7.x engine).

I found that this works for my need:

`var _pspi = new PowerShellProcessInstance(new Version("5.1"), null, ScriptBlock.Create("Add-PsSnapin Microsoft.SharePoint.PowerShell"), false);
var r = RunspaceFactory.CreateOutOfProcessRunspace(TypeTable.LoadDefaultTypeFiles(), _pspi);
r.ThreadOptions = PSThreadOptions.ReuseThread;

var pws = PowerShell.Create();
pws.Runspace = _r;

r.Open();`

at this moment the new Powershell.exe process is spawned, and I can see it from the Process List.
Then I Invoke the Sharepoint script that I have to execute, and the script is run without any issue.

When the script terminates, I need to **close this powershell environment** (while the program continue to run),
I tried every kind of method to close the Powershell, Runspace, Process but it remaisn alive in the system.
A next call to the same function creates a brand new Powershell process... filling all my system memory.

I tried calling:

`_pspi.Process.Close()
_pspi.Process.CloseMainWindow()
r.Close()
r.Dispose()
pws.Dispose()`

All this methods don't fail, but don't terminates the powershell.exe process.

Do I miss something ?
Thanks in advance to all.

### Expected behavior

```console
The powershell.exe process should terminate when the Runspace is disposed or closed or when the Powershell opject is disposed.
```

### Actual behavior

```console
The powershell.exe process never terminates.
```

### Environment data

```powershell
N/A (I don't execute commands directly in a powershell, but through C# code)
```

**Comment by SirDester** [comment]
I add that the only way i found to terminate the powershell process is to call:

```_pspi.Process.Kill()```

but I really don't like this solution

**Comment by rhubarb-geek-nz** [comment]
I agree that calling [TerminateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess) is the absolute last resort.

Are you able send the process a final command to politely exit using [ExitProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess) or similar?

**Comment by SirDester** [comment]
Yes, I also tried invoking "Exit-Process" after the real command (both in the same script command after a ";" and in a separate script command), the invoke completed successfully put the powershell is not closed.

**Comment by SirDester** [comment]
> I agree that calling [TerminateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess) is the absolute last resort.
> 
> Are you able send the process a final command to politely exit using [ExitProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess) or similar, so that executes within the process you want to end? Other low-level alternatives include [GenerateConsoleCtrlEvent](https://learn.microsoft.com/en-us/windows/console/generateconsolectrlevent)

Hello,
after days of digging I finally found the solution (with the help of a good developer :P) and it came debugging all the Powershell SDK calls.

What I (and we) thought it was a "last resort" solution is exactly the solution implemented in the SDK... the Powershell process il **killed** during the Dispose method of the [PowerShellProcessInstance](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/hostifaces/PowerShellProcessInstance.cs) if the process is not exited yet (and I didn't find a way to make the process exit himself); that was the only thing I didn't try to dispose at the beginning.

Anyway I think there's a semantic issue and a lack of documentation.

As the PowerShellProcessInstance object cannot be re-used in a second call of RunspaceFactory.CreateOutOfProcess (i tried and it throws errors) I think that the object should disposed in some way when you dispose the Runspace, because it's strictly related to it.

And even if it's not possible (for some other reasons I can't see) to dispose it it should be told in the documentation that the object need to be Disposed to close (kill) the Powershell process when it's used inside the CreateOutOfProcess method and the Runspace is opened (ahd so the process is started).
